File: src/main/go/g0101_0200/s0101_symmetric_tree/solution.go
package s0101_symmetric_tree

// #Easy #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Breadth_First_Search
// #Tree #Binary_Tree #Data_Structure_I_Day_11_Tree #Level_2_Day_15_Tree
// #Big_O_Time_O(N)_Space_O(log(N)) #2024_03_15_Time_0_ms_(100.00%)_Space_3_MB_(8.89%)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func helper(left *TreeNode, right *TreeNode) bool {
	if left == nil && right == nil {
		return true
	}

	if left == nil || right == nil || left.Val != right.Val {
		return false
	}

	leftRes := helper(left.Left, right.Right)
	rightRes := helper(left.Right, right.Left)
	return leftRes && rightRes
}

func isSymmetric(root *TreeNode) bool {
	return helper(root.Left, root.Right)
}

File: src/main/js/g0101_0200/s0101_symmetric_tree/solution.js
// #Easy #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Breadth_First_Search
// #Tree #Binary_Tree #Data_Structure_I_Day_11_Tree #Level_2_Day_15_Tree
// #Top_Interview_150_Binary_Tree_General #Big_O_Time_O(N)_Space_O(log(N))
// #2024_12_12_Time_0_ms_(100.00%)_Space_53.7_MB_(22.50%)

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function isSymmetric(root) {
    if (!root) {
        return true
    }
    return helper(root.left, root.right)
};

var helper = function(leftNode, rightNode) {
    if (!leftNode || !rightNode) {
        return leftNode === null && rightNode === null
    }
    if (leftNode.val !== rightNode.val) {
        return false
    }
    return helper(leftNode.left, rightNode.right) && helper(leftNode.right, rightNode.left)
};

export { isSymmetric }
